### ISSUE TITLE
✨ How to connect containerized buildkitd remotely without connecting to docker daemon?

### ISSUE THREAD

**Issue by vgjm** [dagger/dagger#2057]
### What are you trying to do?

Expose my buildkitd container through TCP or SSH without expose my docker daemon.

### Why is this important to you?

I want to start a buildkitd container for each user so they can use my machine to execute their dagger plan.
After they executing their dagger plan, I will delete that particular container.
At the same time, I can't share my docker daemon with them because I don't want them to start containers other than buildkitd on my machine.
By the way, I don't want them to connect other user's buildkitd container.
Basically, I want to provide an executing environment(buildkitd) for other users and keep a good isolation property.

**Comment by gerhard** [comment]
I can see three problems wrapped in this issue:

1. Connectivity
2. Access
3. Tenancy


### 1. Connectivity
If a buildkitd container can be network addressed from outside the host where it's running, then it doesn't matter whether it's running in Docker, or some other container runtime.

### 2. Access
Access control will depend on how the buildkitd container is addressable. If it's SSH, then I would expect there to be SSH keys. If it's TCP, then you will most likely have some VPN-based protocol, like Wireguard or Tailscale.

### 3. Tenancy

You don't want to allow a single authenticated user access to all buildkitd instances. I also imagine that you don't want these buildkitd sharing the same resources (noisy neighbour problems). Both can be solved with a good multi-tenancy model.

My first hunch would be to run one Buildkitd instance per [Firecracker VM](https://firecracker-microvm.github.io/), so access to one Buildkitd instance would constrain the client to that single context.

A pool of VMs that is in constant rotation might also work. GitHub Actions has the same model with GitHub Runners (one job = VM = one Docker Engine = one context). While there are other multi-tenancy models that would meet your requirements, I don't think that sharing a single Docker Engine is one of them.

---

I will close this issue since this is a follow-up to this morning's #help discussion with @grouville & @dolanor. This was meant to make it visible - thank you for opening it @vgjm!

<img width="767" alt="image" src="https://user-images.githubusercontent.com/3342/161991570-db93838a-f15b-4eef-a18f-791c7b8d418a.png">
